carbon-design-system/carbon-addons-iot-react
# feat(dashboardeditor): add new render callbacks to allow custom renders of the edit form (and card render) without forcing the Card JSON to change

Closes #

**Summary**
While it was nice to have custom renderEditContent props on the Card and renderCardPreview prop on the DashboardEditor, each had a couple limitations that kept it from being useful. 

The DashboardEditor:renderCardPreview prop was useful because we could dynamically change the rendered card in the DashboardEditor without changing the card JSON. However, it forced us to reproduce all the current DashboardEditorCardRenderer code (with our minor customizations) because DashboardEditorCardRenderer did not expose its normal render behavior in a way that could be overridden.

The Card.renderEditContent prop was nice because we could dynamically add additional form elements to the CardEditor for that Card.  The problem is that we had to do this in the JSON that was passed to the DashboardEditor and then keep removing it from the onChange methods of the Card.  We needed a way to get these form elements to the CardEditForm WITHOUT changing the underlying Card JON.


**Change List (commits, features, bugs, etc)**

- feat(CardEditor): add new custom render callback to the cardeditor allowing additional changes to be made to the card props (to add custom editor elements) WITHOUT requiring it to be passed in the main card JSON.  This allows us to keep the Card JSON focused on the DashboardGrid version of the props, and not add any additional editor props full time to the Card JSON.
- feat(DashboardEditor): add new custom render callback to allow you to change a card JSON dynamically (see above)
- refactor(DashboardEditorCardRenderer): move all the render methods to the DashboardEditorDefaultCardRenderer so it can be reused
- - fix(DashboardEditorCardRenderer): remove the getValidDataItems and dataItems props (and memoization), notice by searching that they are never being used in the Cards downstream, they're only used in the CardEditor
- feat(DashboardEditorDefaultCardRenderer): expose the new DashboardEditorDefaultCardRenderer so downstream components can use it but choose to override it in certain cases. See the new storybook story for an example

**Acceptance Test (how to verify the PR)**

- Verify the new story here: http://127.0.0.1:3000/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-cardeditor--dynamically-add-special-content-to-card-edit-form
 It should add the same kind of EditForm extension that this story adds (http://127.0.0.1:3000/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-cardeditor--table-card-with-custom) but it's doing it dynamically
- Verify that the http://127.0.0.1:3000/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-dashboardeditor--custom-card-preview-renderer story still allows you to render custom card types, but it ALSO allows you to extend the render of existing Card Types by reusing the new component DashboardEditorDefaultCardRenderer

**Regression Test (how to make sure this PR doesn't break old functionality)**

- Verify the existing DashboardEditor/CardEditForm stories still work correctly



**Things to look for during review**
